File: benchmark/template/programs/java/micro/mb02.java
// MB-02 - Factorial: recompute k! mod M from scratch for every k.
class Main {

    private static final long M = 1000003L;
    private static final int N = 20000;

    /** The whole workload body (methodology 06 section 5.2). */
    static long workload() {
        long total = 0;
        for (int k = 1; k <= N; k++) {
            long f = 1;
            for (int j = 2; j <= k; j++) {
                f = (f * j) % M;
            }
            total = (total + f) % M;
        }
        return total;
    }

    public static void main(String[] args) {
        String mode = args.length > 0 ? args[0] : "once";
        int k = args.length > 1 ? Integer.parseInt(args[1]) : 7;

        long total;
        if (mode.equals("steady")) {
            total = 0;
            for (int i = 0; i < k; i++) {
                long t0 = System.nanoTime(); // section 5.2 monotonic clock for Java
                total = workload();
                long elapsed = System.nanoTime() - t0;
                System.out.println("ITER " + i + " " + elapsed);
                System.out.flush(); // section 5.2: flushed immediately
            }
        } else {
            total = workload();
        }

        System.out.println("MB02 " + total);
    }
}

File: benchmark/template/programs/swift/micro/mb02.swift
// MB-02 - Factorial: recompute k! mod M from scratch for every k.
#if canImport(Darwin)
import Darwin  // fflush(stdout) for the section 5.2 ITER lines
#else
import Glibc   // the same fflush(stdout) on Linux, where the benchmark runs
#endif

/// The whole workload body (methodology 06 section 4, MB-02).
func workload() -> String {
    let m = 1000003
    let n = 20000

    var total = 0
    for k in 1...n {
        var f = 1
        // Half-open range: empty at k == 1, which is the pinned behaviour.
        for j in 2..<(k + 1) {
            f = (f * j) % m
        }
        total = (total + f) % m
    }

    return "MB02 \(total)"
}

// Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
func main() {
    let args = CommandLine.arguments
    let mode = args.count > 1 ? args[1] : "once"
    var line: String
    if mode == "steady" {
        var iterations = args.count > 2 ? (Int(args[2]) ?? 7) : 7
        if iterations < 1 {
            iterations = 7
        }
        let clock = ContinuousClock()
        line = ""
        for k in 0..<iterations {
            let elapsed = clock.measure {
                line = workload()
            }
            let parts = elapsed.components
            let ns = parts.seconds * 1_000_000_000 + parts.attoseconds / 1_000_000_000
            print("ITER \(k) \(ns)")
            fflush(stdout)  // section 5.2: each ITER line is flushed immediately
        }
    } else {
        line = workload()
    }
    print(line)
}

main()
